Readme.md

# queue
[![CI](https://github.com/jldec/node-queue4/workflows/CI/badge.svg)](https://github.com/jldec/node-queue4/actions)

Task (function) queue with concurrency / timeout control.

## This is a fork of [tj/node-queue3](https://github.com/tj/node-queue3)
I was doing maintenance on some libraries and ran into vulnerabilities reported by npm because of old versions of dependencies in [queue3](https://github.com/tj/node-queue3).

This fork:
- bumps versions of debug and superagent
- migrates the tests to tape for simplicity and ease of debugging
- updates the example to work with the new superagent api
- removes the Makefile - to run tests do `npm test`

There have been no changes to the module code in index.js.

- The version number in pacakage.json was bumped to 1.0.4 since the original [queue3 on npm](https://www.npmjs.com/package/queue3) reports 1.0.3.
- July 13, 2019: bumped to 1.0.5 with updated deps and example, and CI
- Feb 16, 2020: bumped to 1.0.6 with updated deps and CI on github
- Apr 25, 2020: bumped to 1.0.7 with updated deps
- Jul 26, 2020: bumped to 1.0.8 with updated deps
- Nov 8, 2020: bumped to 1.0.9 with updated deps

## Installation

    $ npm install queue4

## Example

```js

var request = require('superagent');
var Queue = require('./');
var q = new Queue({ concurrency: 3, timeout: 1000 });

var urls = [
  'https://google.com',
  'https://yahoo.com',
  'https://ign.com',
  'https://msn.com',
  'https://hotmail.com',
  'https://cloudup.com',
  'https://learnboost.com'
];

var id = setInterval(function(){
  urls.forEach(function(url){
    q.push(function(fn){
      console.log('%s', url);
      request.get(url, function(err, res){
        console.log('%s -> %s %s',
          url,
          (res && res.status) || 'no response',
          (err && err.message) || 'OK');
        fn();
      });
    });
  });
}, 5000);

var tid = setInterval(function(){
  console.log('%s queued', q.length);
}, 1000);

setTimeout(function(){
  console.log('shutting down');
  clearInterval(id);
  clearInterval(tid);
}, 15000);
```

## License

  MIT


History.md
More recently
==================

* only maintenance releases

1.0.4 / 2019-01-02
==================

 * fork to https://github.com/jldec/node-queue4
 * bump versions of debug and superagent
 * migrate tests to tape for simplicity
 * update the example to work with the more recent superagent
 * release as queue4 on npm

1.0.3 / 2014-06-24
==================

 * no commits or log on github

1.0.2 / 2013-06-26 
==================

 * fix timeout decoration

1.0.1 / 2013-05-22 
==================

 * add better test and timeout() function decorator
 * fix timeout pending reset


test/push.js
var test = require('tape');
var Queue = require('..');

test('should process jobs in order', function(t){
  var q = new Queue;
  var calls = [];

  q.push(function(fn){
    calls.push('one');
    setTimeout(function(){
      calls.push('two');
      fn();
    }, 100);
  });


  q.push(function(fn){
    calls.push('three');
    setTimeout(function(){
      calls.push('four');
      fn();
    }, 100);
  });

  q.push(function(fn){
    calls.push('five');
    setTimeout(function(){
      fn();
      t.deepEquals(calls, ['one', 'two', 'three', 'four', 'five'], 'correct calls');
      t.end();
    }, 100);
  });
});

test('should support .concurrency', function(t){
  var q = new Queue({ concurrency: 5, timeout: 2000 });
  var calls = [];

  q.push(function(fn){
    calls.push('one');
    setTimeout(function(){
      calls.push('two');
      fn();
    }, 100);
  });


  q.push(function(fn){
    calls.push('three');
    setTimeout(function(){
      calls.push('four');
      fn();
    }, 100);
  });

  q.push(function(fn){
    calls.push('five');
    setTimeout(function(){
      fn();
      t.equals(calls.length, 5, 'calls.length');
      t.ok(calls.indexOf('one') !== -1, 'calls includes one');
      t.ok(calls.indexOf('two') !== -1, 'calls includes two');
      t.ok(calls.indexOf('three') !== -1, 'calls includes three');
      t.ok(calls.indexOf('four') !== -1, 'calls includes four');
      t.ok(calls.indexOf('five') !== -1, 'calls includes five');
      t.end();
    }, 100);
  });
});



test/timeout.js
var test = require('tape');
var Queue = require('..');

test('should timeout jobs', function(t){
  var q = new Queue({ timeout: 1000 });
  // var calls = [];

  q.push(function(fn){
    setTimeout(function(){
      fn();
    }, 5000);
  }, function(err){
    t.equal(err.message, 'Timeout of 1000ms exceeded', 'Timeout of 1000ms exceeded');
    t.end();
  });
});

test('should run subsequent jobs', function(t){
  var q = new Queue({ timeout: 500 });
  var calls = [];

  q.push(function(done){
    setTimeout(function(){
      calls.push(1);
      done();
    }, 5000);
  }, function(err){
    t.assert(err.timeout, '1 should timeout');
  });

  q.push(function(done){
    setTimeout(function(){
      calls.push(2);
      done();
    }, 5000);
  }, function(err){
    t.assert(err.timeout, '2 should timeout');
  });

  q.push(function(done){
    setTimeout(function(){
      calls.push(3);
      done();
    }, 100);
  }, function(err){
    t.assert(!err, '3 should not timeout');
  });

  q.push(function(done){
    setTimeout(function(){
      calls.push(4);
      done();
    }, 100);
  }, function(err){
    t.assert(!err, '4 should not timeout');
  });

  q.push(function(){
    t.deepEquals(calls, [3,4], 'include correct calls');
    t.end();
  });
});


index.js

/**
 * Module dependencies.
 */

var Emitter = require('events').EventEmitter;
var debug = require('debug')('queue');

/**
 * Expose `Queue`.
 */

module.exports = Queue;

/**
 * Initialize a `Queue` with the given options:
 *
 *  - `concurrency` [1]
 *  - `timeout` [0]
 *
 * @param {Object} options
 * @api public
 */

function Queue(options) {
  options = options || {};
  this.timeout = options.timeout || 0;
  this.concurrency = options.concurrency || 1;
  this.pending = 0;
  this.jobs = [];
  debug('timeout %s', this.timeout);
  debug('concurrency %s', this.concurrency);
}

/**
 * Inherit from `Emitter.prototype`.
 */

Queue.prototype.__proto__ = Emitter.prototype;

/**
 * Return queue length.
 *
 * @return {Number}
 * @api public
 */

Queue.prototype.__defineGetter__('length', function(){
  return this.pending + this.jobs.length;
});

/**
 * Queue `fn` for execution.
 *
 * @param {Function} fn
 * @param {Function} [cb]
 * @api public
 */

Queue.prototype.push = function(fn, cb){
  debug('enqueue');
  this.jobs.push([fn, cb]);
  process.nextTick(this.run.bind(this));
};

/**
 * Run jobs at the specified concurrency.
 *
 * @api private
 */

Queue.prototype.run = function(){
  while (this.pending < this.concurrency) {
    var job = this.jobs.shift();
    if (!job) break;
    this.exec(job);
  }
};

/**
 * Execute `job`.
 *
 * @param {Array} job
 * @api private
 */

Queue.prototype.exec = function(job){
  var self = this;
  var ms = this.timeout;

  debug('process');
  var fn = job[0];
  var cb = job[1];

  if (ms) fn = timeout(fn, ms);

  this.pending++;
  fn(function(err, res){
    cb && cb(err, res);
    self.pending--;
    self.run();
  });
};

/**
 * Decorate `fn` with a timeout of `ms`.
 *
 * @param {Function} fn
 * @param {Function} ms
 * @return {Function}
 * @api private
 */

function timeout(fn, ms) {
  return function(cb){
    var done;

    var id = setTimeout(function(){
      done = true;
      var err = new Error('Timeout of ' + ms + 'ms exceeded');
      err.timeout = timeout;
      cb(err);
    }, ms);

    fn(function(err, res){
      if (done) return;
      clearTimeout(id);
      cb(err, res);
    });
  };
}


example.js

var request = require('superagent');
var Queue = require('./');
var q = new Queue({ concurrency: 3, timeout: 1000 });

var urls = [
  'https://google.com',
  'https://yahoo.com',
  'https://ign.com',
  'https://msn.com',
  'https://hotmail.com',
  'https://cloudup.com',
  'https://learnboost.com'
];

var id = setInterval(function(){
  urls.forEach(function(url){
    q.push(function(fn){
      console.log('%s', url);
      request.get(url, function(err, res){
        console.log('%s -> %s %s',
          url,
          (res && res.status) || 'no response',
          (err && err.message) || 'OK');
        fn();
      });
    });
  });
}, 5000);

var tid = setInterval(function(){
  console.log('%s queued', q.length);
}, 1000);

setTimeout(function(){
  console.log('shutting down');
  clearInterval(id);
  clearInterval(tid);
}, 15000);
